demo/example_datum.gd

```diff
@@ -0,0 +1,92 @@
+class_name ExampleDatum
+extends Object
+
+# various example data associated with this type
+var _void: VoidData
+var _msg: String
+var _active: bool
+var _int: int
+var _big_int: BigInt
+var _extra_data: Dictionary # assumes that all keys are strings
+
+func _init(
+	void_data := VoidData.new(),
+	msg := "test",
+	active := true,
+	i := 24224,
+	b := BigInt.from_str("-123123123123123123123123123123123").value,
+	extra_data := {
+		"xyz": BigInt.from_int(445)
+	}
+) -> void:
+	_void = void_data
+	_msg = msg
+	_active = active
+	_int = i
+	_big_int = b
+	_extra_data = extra_data
+
+func to_data(strict := false) -> Variant:
+	if strict:
+		var extra_data_encoded := {}
+		for key: String in _extra_data:
+			assert(typeof(key) == TYPE_STRING)
+			var key_encoded := key.to_utf8_buffer()
+			extra_data_encoded[key_encoded] = _extra_data[key]
+			
+		return Constr.new(
+			BigInt.from_int(1),
+			[
+				_void.to_data(strict),
+				_msg.to_utf8_buffer() as Variant,
+				BoolData.new(_active).to_data(strict),
+				BigInt.from_int(_int), 
+				_big_int,
+				extra_data_encoded as Variant
+			]
+		)
+		
+	return Constr.new(
+		BigInt.from_int(1),
+		[
+			_void,
+			_msg,
+			_active,
+			_int, 
+			_big_int,
+			_extra_data
+		]
+	)
+
+static func from_data(v: Variant) -> ExampleDatum:
```

Comment:
I agree 100%.

I think we could leverage Godot's reflection capabilities ([like this function](https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-method-get-property-list)) a generic PlutusData encoding/decoding for all Objects. The encoding could be like this:

* `Object` <-> `Constr 0 [list of fields]` (where each property of the object corresponds with a field inside `Constr`)
* `Array` <-> Plutus list
* `Dictionary<String>` <-> Plutus association list
* `bool` <-> Plutus booleans (this can be a problem, since representation of booleans overlap with objects)
* `int` <-> BigInt
* `String` <-> ByteString

Of course, just as in the current implementation, we can be strict or not with the types (i.e: require a BigInt instead of an int).

The function implementing this encoding/decoding could be provided with the library, so a user can later simply implement `to_data` for their own class like this:

```
func to_data(strict := false) -> Variant:
   return generic_to_data(self)
```

It would be the user's responsibility to make sure that all properties in the class have a `to_data` implementation.